pkg/controllers/dnsrecord/dnsrecord_controller.go

```diff
@@ -0,0 +1,336 @@
+	"github.com/google/go-cmp/cmp/cmpopts"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	utilclock "k8s.io/utils/clock"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	v1 "github.com/Kuadrant/multi-cluster-traffic-controller/pkg/apis/v1"
+)
+
+type ConditionStatus string
+
+const (
+	DNSRecordFinalizer = "kuadrant.io/dns-record"
+
+	ConditionTrue    ConditionStatus = "True"
+	ConditionFalse   ConditionStatus = "False"
+	ConditionUnknown ConditionStatus = "Unknown"
+)
+
+type DNSRecordReconcilerConfig struct {
+	DNSProvider string
+}
+
+// DNSRecordReconciler reconciles a DNSRecord object
+type DNSRecordReconciler struct {
+	client.Client
+	Scheme           *runtime.Scheme
+	ReconcilerConfig DNSRecordReconcilerConfig
+	DNSProvider      dns.Provider
+	DNSZones         []v1.DNSZone
+}
+
+//+kubebuilder:rbac:groups=kuadrant.io,resources=dnsrecords,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=kuadrant.io,resources=dnsrecords/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=kuadrant.io,resources=dnsrecords/finalizers,verbs=update
+
+func (r *DNSRecordReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	previous := &v1.DNSRecord{}
+	err := r.Client.Get(ctx, client.ObjectKey{Namespace: req.Namespace, Name: req.Name}, previous)
+	if err != nil {
+		if err := client.IgnoreNotFound(err); err == nil {
+			return ctrl.Result{}, nil
+		} else {
+			return ctrl.Result{}, err
+		}
+	}
+	dnsRecord := previous.DeepCopy()
+
+	if dnsRecord.DeletionTimestamp != nil && !dnsRecord.DeletionTimestamp.IsZero() {
+		//ToDo Health Checks
+		//if err := c.reconcileHealthCheckDeletion(ctx, dnsRecord); err != nil {
+		//	return err
+		//}
+
+		if err := r.deleteRecord(dnsRecord); err != nil && !strings.Contains(err.Error(), "was not found") {
+			log.Log.Error(err, "Failed to delete DNSRecord", "record", dnsRecord)
+			return ctrl.Result{}, err
+		}
+		controllerutil.RemoveFinalizer(dnsRecord, DNSRecordFinalizer)
+
+		err = r.Update(ctx, dnsRecord)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if !controllerutil.ContainsFinalizer(dnsRecord, DNSRecordFinalizer) {
+		controllerutil.AddFinalizer(dnsRecord, DNSRecordFinalizer)
+		err = r.Update(ctx, dnsRecord)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	statuses := r.publishRecordToZones(r.DNSZones, dnsRecord)
+	if !dnsZoneStatusSlicesEqual(statuses, dnsRecord.Status.Zones) || dnsRecord.Status.ObservedGeneration != dnsRecord.Generation {
+		dnsRecord.Status.Zones = statuses
+		dnsRecord.Status.ObservedGeneration = dnsRecord.Generation
+	}
+
+	//ToDo Health Checks
+	//if err := c.ReconcileHealthChecks(ctx, dnsRecord); err != nil {
```

Comment:
Is this a copy/paste of commented out code,
or something that needs to be implemented still here?

Reply:
Removed all the health check related ToDos. likely if/when we do add health checks it will be different anyway